File: api-rs/src/service/tag_service.rs
use crate::errors::{bad_request, ApiResult};
use crate::model::post::PostRow;
use crate::model::tag::{Tag, TagWithPostCount};
use chrono::Utc;
use sqlx::{query, query_as, Sqlite, SqlitePool, Transaction};
use std::cmp::Reverse;

impl Tag {
    pub async fn get_count(pool: &SqlitePool) -> ApiResult<i64> {
        let count = query!(
            "SELECT COUNT(*) as count FROM tags"
        ).fetch_one(pool).await?.count;

        Ok(count)
    }

    pub async fn get_all_with_post_count(pool: &SqlitePool) -> ApiResult<Vec<TagWithPostCount>> {
        let tags = query_as!(
            TagWithPostCount,
            r#"
            WITH tag_posts AS (
                SELECT t.name AS tag_name, p.id AS post_id
                FROM tags t
                JOIN tag_post_assoc tpa ON t.id = tpa.tag_id
                JOIN posts p ON tpa.post_id = p.id
                WHERE p.deleted_at IS NULL
            )
            SELECT t.name AS name, 
                   t.sticky AS sticky,
                   COUNT(DISTINCT tp.post_id) AS post_count
            FROM tags t
            LEFT JOIN tag_posts tp ON tp.tag_name = t.name OR tp.tag_name LIKE (t.name || '/%')
            GROUP BY t.name
            "#
        ).fetch_all(pool).await?;

        Ok(tags)
    }

    // It will be useful in tests
    #[allow(dead_code)]
    pub async fn get_posts(pool: &SqlitePool, name: &str) -> ApiResult<Vec<PostRow>> {
        let name_like = format!("{}/%", name);
        let posts = query_as!(
            PostRow,
            r#"
            SELECT p.*
            FROM posts p
            WHERE EXISTS (
                SELECT 1
                FROM tags t
                JOIN tag_post_assoc tp ON t.id = tp.tag_id
                WHERE tp.post_id = p.id
                AND (t.name = ? OR t.name LIKE ?))
            AND p.deleted_at IS NULL
            "#,
            name,
            name_like,
        ).fetch_all(pool).await?;

        Ok(posts)
    }

    pub async fn find_or_create(tx: &mut Transaction<'_, Sqlite>, name: &str) -> ApiResult<Tag> {
        let tag = if let Some(tag) = Tag::find_by_name(tx, &name).await? {
            tag
        } else {
            Tag::create(tx, &name).await?
        };
        Ok(tag)
    }

    pub async fn stick(pool: &SqlitePool, name: &str, sticky: bool) -> ApiResult<()> {
        let now = Utc::now().timestamp_millis();

        sqlx::query!(
            r#"
            INSERT INTO tags (name, sticky, created_at, updated_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(name) DO UPDATE SET
                sticky = excluded.sticky,
                updated_at = excluded.updated_at
            "#,
            name,
            sticky,
            now,
            now,
        ).execute(pool).await?;

        Ok(())
    }

    pub async fn delete_associated_posts(pool: &SqlitePool, name: &str) -> ApiResult<()> {
        let now = Utc::now().timestamp_millis();
        let name_like = format!("{}/%", name);

        sqlx::query!(
            r#"
            UPDATE posts
            SET deleted_at = ?1
            WHERE id IN (
                SELECT post_id
                FROM tag_post_assoc
                WHERE tag_id IN (
                    SELECT id
                    FROM tags
                    WHERE name = ?2 OR name LIKE ?3
                )
            )
            "#,
            now,
            name,
            name_like
        ).execute(pool).await?;

        Ok(())
    }

    async fn find_by_name(tx: &mut Transaction<'_, Sqlite>, name: &str) -> ApiResult<Option<Self>> {
        let tag = query_as!(
            Tag,
            "SELECT * FROM tags WHERE name = ?",
            name,
        ).fetch_optional(&mut **tx).await?;

        Ok(tag)
    }

    async fn create(tx: &mut Transaction<'_, Sqlite>, name: &str) -> ApiResult<Tag> {
        let now = Utc::now().timestamp_millis();

        let id = query!(
            r#"
            INSERT INTO tags (name, sticky, created_at, updated_at)
            VALUES (?, false, ?, ?)
            RETURNING id
            "#,
            name,
            now,
            now
        ).fetch_one(&mut **tx).await?.id;

        Ok(Tag {
            id,
            name: name.to_string(),
            sticky: false,
            created_at: now,
            updated_at: now,
        })
    }

    /// Rename a tag, and if the new tag already exists, merge the tags.
    /// Handles all descendant tags recursively with optimal performance.
    pub async fn rename_or_merge(pool: &SqlitePool, name: &str, new_name: &str) -> ApiResult<()> {
        if name == new_name {
            return Ok(());
        }

        // Check for invalid hierarchy
        if new_name.starts_with(name) && new_name.matches('/').count() > name.matches('/').count() {
            return Err(bad_request(
                &format!(r#"Cannot move "{}" to a subtag of itself "{}""#, name, new_name)
            ));
        }

        let name_like = format!("{}/%", name);

        // Get all affected tags in a single query
        let mut affected_tags = sqlx::query_as!(
            Tag,
            r#"
            SELECT * FROM tags
            WHERE name = ? OR name = ? OR name LIKE ?
            "#,
            name,
            new_name,
            name_like
        ).fetch_all(pool).await?;

        let mut tx = pool.begin().await?;

        // Split into source tag, target tag and descendants
        let source_tag = if let Some(tag) = affected_tags.iter().find(|t| t.name == name) {
            tag
        } else {
            let new_tag = Tag::create(&mut tx, name).await?;
            affected_tags.push(new_tag);
            affected_tags.last().unwrap()
        };

        let target_tag = affected_tags.iter().find(|t| t.name == new_name);

        let mut descendants: Vec<_> = affected_tags
            .iter()
            .filter(|t| t.name != name && t.name != new_name)
            .collect();
        descendants.sort_by_key(|t| Reverse(t.name.matches('/').count()));

        for descendant in descendants {
            let new_descendant_name = descendant.name.replace(name, new_name);
            let target_descendant = Tag::find_by_name(&mut tx, &new_descendant_name).await?;

            if let Some(target_descendant) = target_descendant {
                // Target exists - merge
                Tag::merge(&mut tx, descendant, &target_descendant).await?;
            } else {
                // Target doesn't exist - rename
                Tag::rename(&mut tx, descendant, &new_descendant_name).await?;
            }
        }

        if let Some(target_tag) = target_tag {
            Tag::merge(&mut tx, source_tag, target_tag).await?;
        } else {
            Tag::rename(&mut tx, source_tag, new_name).await?;
        }

        tx.commit().await?;
        Ok(())
    }

    /// Rename a tag and update all related post content.
    async fn rename(
        tx: &mut Transaction<'_, Sqlite>,
        tag: &Tag,
        new_name: &str,
    ) -> ApiResult<()> {
        let now = Utc::now().timestamp_millis();

        // Update tag name
        sqlx::query!(
            r#"
            UPDATE tags
            SET name = ?, updated_at = ?
            WHERE id = ?
            "#,
            new_name,
            now,
            tag.id
        ).execute(&mut **tx).await?;

        let source_pattern = format!(">#{}<", tag.name);
        let target_pattern = format!(">#{}<", new_name);

        // Update post content using a single query
        sqlx::query!(
            r#"
            UPDATE posts
            SET content = REPLACE(content, ?, ?)
            WHERE id IN (
                SELECT post_id
                FROM tag_post_assoc
                WHERE tag_id = ?
            )
            "#,
            source_pattern,
            target_pattern,
            tag.id
        ).execute(&mut **tx).await?;

        Ok(())
    }

    /// Merge one tag into another, updating all related posts
    async fn merge(
        tx: &mut Transaction<'_, Sqlite>,
        source_tag: &Tag,
        target_tag: &Tag,
    ) -> ApiResult<()> {
        // Update post content
        let source_pattern = format!(">#{}<", source_tag.name);
        let target_pattern = format!(">#{}<", target_tag.name);

        sqlx::query!(
            r#"
            UPDATE posts
            SET content = REPLACE(content, ?, ?)
            WHERE id IN (
                SELECT post_id
                FROM tag_post_assoc
                WHERE tag_id = ?
            )
            "#,
            source_pattern,
            target_pattern,
            source_tag.id
        ).execute(&mut **tx).await?;

        // Insert new tag associations (ignoring if they already exist)
        sqlx::query!(
            r#"
            INSERT OR IGNORE INTO tag_post_assoc (post_id, tag_id)
            SELECT post_id, ? as tag_id
            FROM tag_post_assoc
            WHERE tag_id = ?
            "#,
            target_tag.id,
            source_tag.id
        ).execute(&mut **tx).await?;

        // Delete old tag associations
        sqlx::query!(
            r#"
            DELETE FROM tag_post_assoc
            WHERE tag_id = ?
            "#,
            source_tag.id
        ).execute(&mut **tx).await?;

        Ok(())
    }
}

File: api-kt/src/main/kotlin/site/daydream/mote/service/TagService.kt
package site.daydream.mote.service

import site.daydream.mote.exception.BadRequestException
import site.daydream.mote.generated.Tables.POSTS
import site.daydream.mote.generated.Tables.TAGS
import site.daydream.mote.model.Post
import site.daydream.mote.model.Tag
import site.daydream.mote.model.TagWithPostCount
import site.daydream.mote.util.count
import site.daydream.mote.util.replaceFromStart
import org.jooq.DSLContext
import org.jooq.impl.DSL
import org.jooq.impl.DSL.selectOne
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import site.daydream.mote.generated.Tables.TAG_POST_ASSOC as ASSOC


@Service
@Transactional
class TagService(private val dsl: DSLContext) {
    fun findByName(name: String): Tag? =
        dsl.selectFrom(TAGS)
            .where(TAGS.NAME.eq(name))
            .fetchOneIntoClass()

    fun getCount(): Int =
        dsl.fetchCount(TAGS)

    fun getAllWithPostCount(): List<TagWithPostCount> {
        val sql = """
            SELECT t.name, t.sticky,
                (
                    SELECT COUNT(DISTINCT a.post_id)
                    FROM tag_post_assoc a
                    WHERE a.tag_id IN (
                        SELECT id
                        FROM tags
                        WHERE name = t.name
                           OR name LIKE t.name || '/%'
                    )
            ) AS post_count
            FROM tags t;
        """
        return dsl.resultQuery(sql).fetchAllIntoClass()
    }

    fun getPosts(name: String): List<Post> {
        return dsl.selectFrom(POSTS)
            .whereExists(
                selectOne()
                    .from(TAGS)
                    .join(ASSOC).on(TAGS.ID.eq(ASSOC.TAG_ID))
                    .where(ASSOC.POST_ID.eq(POSTS.ID))
                    .and(TAGS.NAME.eq(name).or(TAGS.NAME.startsWith("$name/")))
            )
            .and(POSTS.DELETED_AT.isNull)
            .fetchAllIntoClass()
    }

    fun findOrCreate(name: String): Tag {
        return findByName(name) ?: create(name)
    }

    fun insertOrUpdate(name: String, sticky: Boolean) {
        val now = Instant.now().toEpochMilli()

        dsl.insertInto(TAGS)
            .columns(TAGS.NAME, TAGS.STICKY, TAGS.CREATED_AT, TAGS.UPDATED_AT)
            .values(name, sticky, now, now)
            .onConflict(TAGS.NAME)
            .doUpdate()
            .set(TAGS.STICKY, sticky)
            .set(TAGS.UPDATED_AT, now)
            .execute()
    }

    fun create(name: String): Tag {
        val record = dsl.newRecord(TAGS, Tag(name = name))
        record.store()
        return Tag(id = record.id, name = record.name)
    }

    /**
     * Soft delete all posts under this tag and its descendant tags.
     * Performs the operation in a single transaction with minimal database access.
     */
    fun deleteAssociatedPosts(name: String) {
        val now = Instant.now().toEpochMilli()

        dsl.update(POSTS)
            .set(POSTS.DELETED_AT, now)
            .where(
                POSTS.ID.`in`(
                    dsl.select(ASSOC.POST_ID).from(ASSOC)
                        .where(
                            ASSOC.TAG_ID.`in`(
                                dsl.select(TAGS.ID).from(TAGS)
                                    .where(TAGS.NAME.eq(name).or(TAGS.NAME.startsWith("$name/")))
                            )
                        )
                )
            )
            .execute()
    }

    /**
     * Rename a tag, and if the new tag already exists, merge the tags.
     * Handles all descendant tags recursively with optimal performance.
     */
    fun renameOrMerge(name: String, newName: String) {
        if (name == newName) return
        if (newName.startsWith(name) && newName.count('/') > name.count('/')) {
            throw BadRequestException("""Cannot move "$name" to a subtag of itself "$newName"""")
        }

        // Get all affected tags in a single query
        val affectedTags = dsl
            .selectFrom(TAGS)
            .where(
                TAGS.NAME.eq(name)
                    .or(TAGS.NAME.eq(newName))
                    .or(TAGS.NAME.startsWith("$name/"))
            )
            .fetchAllIntoClass<Tag>()

        // Split into source tag, target tag and descendants
        val sourceTag = affectedTags.find { it.name == name } ?: create(name)
        val targetTag = affectedTags.find { it.name == newName }
        val descendants = affectedTags
            .filter { it.name != name && it.name != newName }
            .sortedByDescending { it.name.count('/') }

        descendants.forEach { descendant ->
            val newDescendantName = descendant.name.replaceFromStart(name, newName)
            val targetDescendant = findByName(newDescendantName)
            if (targetDescendant != null) {
                // Target exists - merge
                merge(descendant, targetDescendant)
            } else {
                // Target doesn't exist - rename
                rename(descendant, newDescendantName)
            }
        }

        if (targetTag != null) {
            merge(sourceTag, targetTag)
        } else {
            rename(sourceTag, newName)
        }
    }

    /**
     * Rename a tag and update all related post content.
     * Performs the operation in a single transaction.
     */
    private fun rename(tag: Tag, newName: String) {
        val oldName = tag.name

        // Update tag name
        dsl.update(TAGS)
            .set(TAGS.NAME, newName)
            .set(TAGS.UPDATED_AT, Instant.now().toEpochMilli())
            .where(TAGS.ID.eq(tag.id))
            .execute()

        // Update post content using a single query
        dsl.update(POSTS)
            .set(
                POSTS.CONTENT,
                DSL.replace(POSTS.CONTENT, ">#$oldName<", ">#$newName<")
            )
            .where(
                POSTS.ID.`in`(
                    dsl.select(ASSOC.POST_ID)
                        .from(ASSOC)
                        .where(ASSOC.TAG_ID.eq(tag.id))
                )
            )
            .execute()
    }

    /**
     * Merge one tag into another, updating all related posts.
     * Performs the operation in a single transaction.
     */
    private fun merge(sourceTag: Tag, targetTag: Tag) {
        // Get all posts associated with the source tag
        val postIds = dsl.select(ASSOC.POST_ID)
            .from(ASSOC)
            .where(ASSOC.TAG_ID.eq(sourceTag.id))
            .fetch(ASSOC.POST_ID)

        if (postIds.isEmpty()) return

        // Update post content
        dsl.update(POSTS)
            .set(
                POSTS.CONTENT,
                DSL.replace(POSTS.CONTENT, ">#${sourceTag.name}<", ">#${targetTag.name}<")
            )
            .where(POSTS.ID.`in`(postIds))
            .execute()

        // Insert new tag associations (ignoring if they already exist)
        dsl.insertInto(ASSOC)
            .columns(ASSOC.POST_ID, ASSOC.TAG_ID)
            .select(
                dsl.select(ASSOC.POST_ID, DSL.value(targetTag.id).`as`("tag_id"))
                    .from(ASSOC)
                    .where(ASSOC.TAG_ID.eq(sourceTag.id))
            )
            // SQLite syntax, for PostgreSQL use `onDuplicateKeyIgnore` instead
            .onConflictDoNothing()
            .execute()

        // Delete old tag associations
        dsl.deleteFrom(ASSOC)
            .where(ASSOC.TAG_ID.eq(sourceTag.id))
            .execute()

        // Delete the source tag itself
        dsl.deleteFrom(TAGS)
            .where(TAGS.ID.eq(sourceTag.id))
            .execute()
    }
}
